### PR TITLE
Adding support for Unreal Engine 5.4

### PR DESCRIPTION
**Pull request by deveshmishrabetide**
- Replaced old Json Function by New
- Added header for Settings along with LocText Fix
- Image functions changed/Added

### CODE REVIEW

File: src/UE5.4/Edgegap/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
+FROM ubuntu
+
+RUN apt-get update
+RUN apt-get install openssh-server sudo jq -y && \
```
Comment (orcist) [inline src/UE5.4/Edgegap/Dockerfile:4]: @deveshmishrabetide are you sure `sudo` should be here?

Comment (deveshmishrabetide) [reply src/UE5.4/Edgegap/Dockerfile:4]: That is what is used by the older versions. I took your changes for those two files : 

https://github.com/edgegap/edgegap-unreal-plugin/blob/main/src/UE5.0/Edgegap/Dockerfile

File: src/UE5.4/Edgegap/StartServer.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+GAME_PORT=$(echo $ARBITRIUM_PORTS_MAPPING | jq '.ports.gameport.external')
```
Comment (orcist) [inline src/UE5.4/Edgegap/StartServer.sh:3]:
```suggestion
GAME_PORT=$(echo $ARBITRIUM_PORTS_MAPPING | jq '.ports.gameport.internal')
```

File: src/UE5.4/Edgegap/Dockerfile (hunk 1 above)
Comment (deveshmishrabetide) [reply src/UE5.4/Edgegap/Dockerfile:4]: To add, it's same on 5.3 too: 

https://github.com/edgegap/edgegap-unreal-plugin/blob/main/src/UE5.3/Edgegap/Dockerfile

Comment (orcist) [reply src/UE5.4/Edgegap/Dockerfile:4]: okay I might open another PR later to remove this, thanks for the note!

File: src/UE5.4/Edgegap/Source/Edgegap/Private/EdgegapSettingsDetails.cpp [hunk 3]
```diff
@@ -0,0 +1,2149 @@
...
+	const FSlateBrush* ContentPrepIcon = nullptr;
+	if (true)
+	{
+		ContentPrepDescription = LOCTEXT("PackagingProjectTaskName", "Packaging project");
+		ContentPrepTaskName = LOCTEXT("PackagingTaskName", "Packaging");
+#if ENGINE_MAJOR_VERSION >= 5 && ENGINE_MINOR_VERSION > 3
+		ContentPrepIcon = FAppStyle::GetBrush(TEXT("MainFrame.PackageProject"));
+#else
+		ContentPrepIcon = FEditorStyle::GetBrush(TEXT("MainFrame.PackageProject"));
+#endif
```
Comment (orcist) [inline src/UE5.4/Edgegap/Source/Edgegap/Private/EdgegapSettingsDetails.cpp:1051]: note: this has been added

File: src/UE5.4/Edgegap/Source/Edgegap/Private/EdgegapSettingsDetails.cpp [hunk 4]
```diff
@@ -0,0 +1,2149 @@
...
+	StartScriptContent = StartScriptContent.Replace(*FString("<PROJECT_NAME>"), FApp::GetProjectName());
+	FFileHelper::SaveStringToFile(StartScriptContent, *NewStartScriptPath);
+
+	FString CommandLine = FString::Printf(TEXT("docker build -t \"%s\" \"%s\""), *_ImageName, *ServerBuildPath);
+	UE_LOG(EdgegapLog, Log, TEXT("%s"), *CommandLine);
+#if ENGINE_MAJOR_VERSION >= 5 && ENGINE_MINOR_VERSION > 3
+	IUCMDHelperModule::Get().CreateUcmdTask(CommandLine, LOCTEXT("DisplayName", "Docker"), LOCTEXT("ContainerizingProjectTaskName", "Containerizing server"), LOCTEXT("ContainerizingTaskName", "Containerizing"), FAppStyle::GetBrush(TEXT("MainFrame.PackageProject")), false, &OnContainerizeCallback);
+#else
+	IUCMDHelperModule::Get().CreateUcmdTask(CommandLine, LOCTEXT("DisplayName", "Docker"), LOCTEXT("ContainerizingProjectTaskName", "Containerizing server"), LOCTEXT("ContainerizingTaskName", "Containerizing"), FEditorStyle::GetBrush(TEXT("MainFrame.PackageProject")), false, &OnContainerizeCallback);
+#endif
```
Comment (orcist) [inline src/UE5.4/Edgegap/Source/Edgegap/Private/EdgegapSettingsDetails.cpp:1222]: note: this has been added

File: src/UE5.4/Edgegap/Source/Edgegap/Private/EdgegapSettingsDetails.cpp [hunk 5]
```diff
@@ -0,0 +1,2149 @@
...
+		return;
+	}
+
+	FString CommandLine = FString::Printf(TEXT("docker image push %s"), *ImageName);
+	UE_LOG(EdgegapLog, Log, TEXT("%s"), *CommandLine);
+#if ENGINE_MAJOR_VERSION >= 5 && ENGINE_MINOR_VERSION > 3
+	IUCMDHelperModule::Get().CreateUcmdTask(CommandLine, LOCTEXT("DisplayName", "Docker"), LOCTEXT("PushContainerProjectTaskName", "Pushing Container"), LOCTEXT("ContainerizingTaskName", "Docker Login"), FAppStyle::GetBrush(TEXT("MainFrame.PackageProject")), false, &OnPushContainerCallback);
+#else
+	IUCMDHelperModule::Get().CreateUcmdTask(CommandLine, LOCTEXT("DisplayName", "Docker"), LOCTEXT("PushContainerProjectTaskName", "Pushing Container"), LOCTEXT("ContainerizingTaskName", "Docker Login"), FEditorStyle::GetBrush(TEXT("MainFrame.PackageProject")), false, &OnPushContainerCallback);
+#endif
```
Comment (orcist) [inline src/UE5.4/Edgegap/Source/Edgegap/Private/EdgegapSettingsDetails.cpp:1244]: note: this has been added

File: src/UE5.4/Edgegap/Source/Edgegap/Private/EdgegapSettingsDetails.cpp [hunk 6]
```diff
@@ -0,0 +1,2149 @@
...
+
+void FEdgegapSettingsDetails::DockerLogin(FString RegistryURL, FString PrivateUsername, FString PrivateToken)
+{
+	FString CommandLine = FString::Printf(TEXT("echo \'%s\' | docker login -u \'%s\' --password-stdin %s"), *PrivateToken, *PrivateUsername, *RegistryURL);
+	UE_LOG(EdgegapLog, Log, TEXT("%s"), *CommandLine);
+#if ENGINE_MAJOR_VERSION >= 5 && ENGINE_MINOR_VERSION > 3
+	IUCMDHelperModule::Get().CreateUcmdTask(CommandLine, LOCTEXT("DisplayName", "Docker"), LOCTEXT("DockerLoginProjectTaskName", "Logging into Registry"), LOCTEXT("ContainerizingTaskName", "Docker Login"), FAppStyle::GetBrush(TEXT("MainFrame.PackageProject")), true, &OnDockerLoginCallback);
+#else
+	IUCMDHelperModule::Get().CreateUcmdTask(CommandLine, LOCTEXT("DisplayName", "Docker"), LOCTEXT("DockerLoginProjectTaskName", "Logging into Registry"), LOCTEXT("ContainerizingTaskName", "Docker Login"), FEditorStyle::GetBrush(TEXT("MainFrame.PackageProject")), true, &OnDockerLoginCallback);
+#endif
```
Comment (orcist) [inline src/UE5.4/Edgegap/Source/Edgegap/Private/EdgegapSettingsDetails.cpp:1255]: note: this has been added

File: src/UE5.4/Edgegap/Source/Edgegap/Private/EdgegapSettingsDetails.cpp [hunk 7]
```diff
@@ -0,0 +1,2149 @@
...
+			FString Response = ResponsePtr->GetContentAsString();
+
+			TSharedPtr<FJsonValue> JsonValue;
+			TSharedRef<TJsonReader<>> Reader = TJsonReaderFactory<>::Create(Response);
+
+			if (ResponseCode == 204 || FJsonSerializer::Deserialize(Reader, JsonValue))
+			{
+				if (ResponseCode != 204 && JsonValue)
+				{
+					if (JsonValue->AsObject()->HasField(UTF8_TO_TCHAR("message")))
```
Comment (orcist) [inline src/UE5.4/Edgegap/Source/Edgegap/Private/EdgegapSettingsDetails.cpp:1328]: note: `UTF8_TO_TCHAR` has been added

File: src/UE5.4/Edgegap/Source/Edgegap/Private/EdgegapSettingsDetails.cpp [hunk 8]
```diff
@@ -0,0 +1,2149 @@
...
+					{
+						FString message = JsonValue->AsObject()->GetStringField(UTF8_TO_TCHAR("message"));
```
Comment (orcist) [inline src/UE5.4/Edgegap/Source/Edgegap/Private/EdgegapSettingsDetails.cpp:1330]: note: `UTF8_TO_TCHAR` has been added

File: src/UE5.4/Edgegap/Source/Edgegap/Private/EdgegapSettingsDetails.cpp [hunk 9]
```diff
@@ -0,0 +1,2149 @@
...
+
+	// Set request fields
+	IpifyRequest->SetURL(IpifyIPsURL);
+	IpifyRequest->SetVerb("GET");
+	IpifyRequest->SetHeader(TEXT("User-Agent"), TEXT("X-UnrealEngine-Agent"));
+
+	IpifyRequest->SetHeader("Content-Type", "application/json");
+	IpifyRequest->SetHeader(TEXT("Authorization"), *API_key);
+
+	IpifyRequest->OnProcessRequestComplete().BindLambda([this, AppName, VersionName, InCreateNewDeployment_SBtn, API_key](FHttpRequestPtr RequestRes, FHttpResponsePtr ResponsePtr, bool bWasSuccessful)
```
Comment (orcist) [inline src/UE5.4/Edgegap/Source/Edgegap/Private/EdgegapSettingsDetails.cpp:1773]: note: `FHttpRequestPtr` type changed from `Request` to `RequestRes`

File: src/UE5.4/Edgegap/Source/Edgegap/Private/APIToken/APITokenSettingsCustomization.cpp [hunk 10]
```diff
@@ -0,0 +1,80 @@
+#include "APITokenSettingsCustomization.h"
+#include "APITokenSettings.h"
+#include "DetailLayoutBuilder.h"
+#include "DetailWidgetRow.h"
+#include "EdgegapSettingsDetails.h"
```
Comment (orcist) [inline src/UE5.4/Edgegap/Source/Edgegap/Private/APIToken/APITokenSettingsCustomization.cpp:5]: note: this has been added

File: src/UE5.4/Edgegap/Source/Edgegap/Private/APIToken/APITokenSettingsCustomization.cpp [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+#include "Widgets/Input/SButton.h"
+#include "JsonObjectConverter.h"
+#include "Widgets/SBoxPanel.h"
+#define LOCTEXT_NAMESPACE "APITokenSettingsCustomization"
```
Comment (orcist) [inline src/UE5.4/Edgegap/Source/Edgegap/Private/APIToken/APITokenSettingsCustomization.cpp:9]: note: this has been added

**Comment by orcist** [comment]
@deveshmishrabetide we'll need to bring back the `Binaries` folder, since Unreal determines plugin+engine compatibility by comparing UE's `Engine/Binaries/Win64/UnrealEditor.modules` with the plugin's `Binaries/Win64/UnrealEditor-Win64-DebugGame.modules` IDs (target platform). Without it, Unreal always requires rebuilding the plugin from source.

**Comment by deveshmishrabetide** [comment]
> @deveshmishrabetide we'll need to bring back the `Binaries` folder, since Unreal determines plugin+engine compatibility by comparing UE's `Engine/Binaries/Win64/UnrealEditor.modules` with the plugin's `Binaries/Win64/UnrealEditor-Win64-DebugGame.modules` IDs (target platform). Without it, Unreal always requires rebuilding the plugin from source.

Sure thing! Will do

**Comment by orcist** [comment]
@deveshmishrabetideI've discovered that in UE builds from source, it’s a UUID, e.g. `2f8c7431-93fe-4a34-af4b-c4c63b9c1b3b` for my build of 5.4.4. The UUID is different every time you build UE from source (tested on 2 separate machines), so the plugin will always complain about compatibility. This makes me think having pre-built plugins for UE built from source is actually impossible, and we might as well just keep the source version here and instruct the user on how to rebuild the plugin themselves.

I suppose no need to bother with rebuilding and including the binaries folder for now. 👍

**Comment by orcist** [comment]
Cancelling this PR, see previous comment. We'll provide a detailed guide with a video for building the plugin custom for Unreal built from source. We're discussing developing a Docker Desktop extension as a replacement for the plugin, removing the need for the plugin rebuilding step completely... stay tuned for updates!

**Comment by orcist** [comment]
reopening, even though rebuilding from source is required, this PR includes a few more fixes for 5.4 which we'd like to have

**Pull request merged**